### PR TITLE
Upgrade to elfinder api 2.1 and remove the transport api adapter 

### PR DESCRIPTION
**Pull request by oeway**
This PR get rid of the transport adapter (`elFinderSupportVer1.js`) and implement the new api directly in the backend.

### CODE REVIEW

File: jupyter_elfinder/elfinder.py [hunk 1]
```diff
@@ -391,28 +360,109 @@ def run(
...
-        if "tree" not in self._request and "current" in self._request:
-            cur_dir = self.__find_dir(self._request["current"], None)
+        # try dir
 
-            if not cur_dir:
-                self.http_status_code = 404
-                self.http_header["Content-type"] = "text/html"
-                self.http_response = "File not found"
+        # initialized = len(self._cached_path) > 0
+        if "init" in self._request and self._request["init"]:
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]: Invert this check. Always prefer guard clause when possible. It saves on indentation.
```py
if not self._request.get("info") and not self._request.get("target"):
    self._response["error"] = "Invalid parameters"
    return
...
```

File: jupyter_elfinder/elfinder.py [hunk 2]
```diff
@@ -391,28 +360,109 @@ def run(
...
+            self._response["api"] = 2.1
+        else:
+            if "target" not in self._request:
+                self._response["error"] = "Invalid parameters"
                 return
 
-            cur_file = self.__find(self._request["target"], cur_dir)
+        if "target" in self._request and self._request["target"]:
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```py
target = self._request.get("target")
if target:
    target = self.__find_dir(target)
...
```

File: jupyter_elfinder/elfinder.py [hunk 3]
```diff
@@ -391,28 +360,112 @@ def run(
...
+            target = self.__find_dir(self._request["target"])
+            if not target:
+                self._response["error"] = (
+                    "Invalid parameters: " + self._request["target"]
+                )
+            elif not self.__is_allowed(target, "read"):
+                self._response["error"] = "Access denied"
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]: If we get to here, don't we want to return after this? Should we really continue on error?

File: jupyter_elfinder/elfinder.py [hunk 4]
```diff
@@ -531,9 +563,9 @@ def __mkdir(self) -> None:
             name = self._request["name"]
             name = self.__check_utf8(name)
             target = self._request["target"]
-            path = self.__find_dir(target, None)
-        if "dirs" in self._request:
-            dirs = self._request["dirs"]
+            path = self.__find_dir(target)
+        if "dirs[]" in self._request:
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```py
dirs = self._request.get("dirs[]") or []
```

File: jupyter_elfinder/elfinder.py [hunk 5]
```diff
@@ -941,24 +961,12 @@ def __cwd(self, path: str) -> None:
...
-            info = self.__info(file_path)
-            info["hash"] = self.__hash(file_path)
-            if info["mime"] == "directory":
-                dirs.append(info)
-            else:
-                files.append(info)
+        try:
+            info["dirs"] = any(next(os.walk(path))[1])
+        except StopIteration:
+            info["dirs"] = False
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
            info["dirs"] = None
```

File: jupyter_elfinder/elfinder.py [hunk 6]
```diff
@@ -1060,40 +1071,110 @@ def __info(self, path: str) -> Info:
...
+            all_file_count += file_count
+            all_dir_count += dir_count
+
+        self._response["size"] = all_total_size
+        self._response["fileCnt"] = all_file_count
+        self._response["dirCnt"] = all_dir_count
+        self._response["sizes"] = sizes
+
+    def __ls(self) -> None:
+        if not self._request.get("target"):
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
        target = self._request.get("target")
        if not target:
```

File: jupyter_elfinder/elfinder.py [hunk 7]
```diff
@@ -1060,40 +1071,110 @@ def __info(self, path: str) -> Info:
...
+            self._response["error"] = "Invalid parameters"
+            return
+
+        target = self._request["target"]
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
```

File: jupyter_elfinder/elfinder.py [hunk 8]
```diff
@@ -1060,40 +1071,110 @@ def __info(self, path: str) -> Info:
...
+        intersect = self._request.get("intersect[]")
+
+        path = self.__find(target)
+        if path is None or not os.path.isdir(path):
+            self._response["error"] = "Target directory not found."
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]: We don't end the other error messages with period.
```suggestion
            self._response["error"] = "Target directory not found"
```

File: jupyter_elfinder/elfinder.py [hunk 9]
```diff
@@ -1060,40 +1071,110 @@ def __info(self, path: str) -> Info:
...
+                items[fhash] = fname
+        self._response["list"] = items
+
+    def __tree(self) -> None:
+        """Return directory tree starting from path."""
+        if "target" not in self._request:
+            self._response["error"] = "Invalid parameters"
+            return
+        target = self._request["target"]
+        path = self.__find_dir(target)
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]: `__find_dir` must have a string as argument. Do we need to check for None first?

File: jupyter_elfinder/elfinder.py [hunk 10]
```diff
@@ -1060,40 +1071,110 @@ def __info(self, path: str) -> Info:
...
 
-        # limit the tree depth to 1
-        if depth < 1 and self.__is_allowed(path, "read"):
+        if path is None or not os.path.isdir(path):
+            self._response["error"] = "Directory not found."
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
            self._response["error"] = "Directory not found"
```

File: jupyter_elfinder/elfinder.py [hunk 11]
```diff
@@ -1060,40 +1071,110 @@ def __info(self, path: str) -> Info:
...
+            return
+
+        if os.path.islink(path):
+            path = self.__read_link(path)
+            if path is None:
+                self._response["error"] = "Directory (link) not found."
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
                self._response["error"] = "Directory (link) not found"
```

File: jupyter_elfinder/elfinder.py [hunk 12]
```diff
@@ -1261,22 +1333,15 @@ def __dim(self) -> None:
...
             return
 
-        cur_dir = self.__find_dir(self._request["current"], None)
-
-        if not cur_dir:
-            self._response["error"] = "File not found"
-            return
-
-        cur_file = self.__find(self._request["target"], cur_dir)
+        cur_file = self.__find(self._request["target"])
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
        cur_file = self.__find(target)
```

File: jupyter_elfinder/elfinder.py [hunk 13]
```diff
@@ -1360,34 +1423,26 @@ def __archive(self) -> None:
...
+        if not self._options["archivers"]["extract"] or "target" not in self._request:
             self._response["error"] = "Invalid parameters"
             return
-
-        cur_dir = self.__find_dir(self._request["current"], None)
-
-        if not cur_dir:
+        target = self._request["target"]
+        makedir = self._request.get("makedir")
+        cur_file = self.__find(target)
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]: `target` must be a string here. Do we need to check for that?

File: jupyter_elfinder/elfinder.py (hunk 3 above)
Comment (oeway) [reply jupyter_elfinder/elfinder.py]: You are right, we should return.

File: jupyter_elfinder/elfinder.py (hunk 5 above)
Comment (oeway) [reply jupyter_elfinder/elfinder.py]: Right, I just noticed that in the spec, it `dirs` should be a Number, I changed this in a later commit.

File: jupyter_elfinder/elfinder.py (hunk 13 above)
Comment (oeway) [reply jupyter_elfinder/elfinder.py]: All fixed in a later commit

File: jupyter_elfinder/elfinder.py [hunk 14]
```diff
@@ -1060,40 +1073,115 @@ def __info(self, path: str) -> Info:
...
+
+    def __tree(self) -> None:
+        """Return directory tree starting from path."""
+        if "target" not in self._request:
+            self._response["error"] = "Invalid parameters"
+            return
+        target = self._request["target"]
+        if not target:
+            self._response["error"] = "Invalid parameters"
+            return
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
        target = self._request.get("target")
        if not target:
            self._response["error"] = "Invalid parameters"
            return
```

File: jupyter_elfinder/elfinder.py [hunk 15]
```diff
@@ -1060,40 +1083,115 @@ def __info(self, path: str) -> Info:
...
+    def __ls(self) -> None:
+        if not self._request.get("target"):
+            self._response["error"] = "Invalid parameters"
+            return
+
+        target = self._request["target"]
+        intersect = self._request.get("intersect[]")
+        if not target:
+            self._response["error"] = "Invalid parameters"
+            return
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
```

File: jupyter_elfinder/elfinder.py [hunk 16]
```diff
@@ -1060,40 +1083,115 @@ def __info(self, path: str) -> Info:
...
+            self._response["error"] = "Directory not found"
+            return
+
+        if os.path.islink(path):
+            path = self.__read_link(path)
+            if path is None:
+                self._response["error"] = "Directory (link) not found"
+                return
+
+        if self.__is_allowed(path, "read"):
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]: Then we can outdent below.
```suggestion
        if not self.__is_allowed(path, "read"):
            self._response["error"] = "Access denied"
            return
```

File: jupyter_elfinder/elfinder.py [hunk 17]
```diff
@@ -1060,40 +1083,115 @@ def __info(self, path: str) -> Info:
...
                     os.path.isdir(dir_path)
                     and not os.path.islink(dir_path)
                     and self.__is_accepted(directory)
                 ):
-                    dirs.append(self.__tree(dir_path, depth + 1))
-        return tree
+                    tree.append(self.__info(dir_path))
+            self._response["tree"] = tree
+        else:
+            self._response["error"] = "Access denied"
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
```

File: jupyter_elfinder/elfinder.py [hunk 18]
```diff
@@ -1360,34 +1440,29 @@ def __archive(self) -> None:
...
+            self._response["added"] = [self.__info(archive_path)]
         else:
             self._response["error"] = "Unable to create archive"
 
     def __extract(self) -> None:
         """Uncompress archive."""
-        current = self._request.get("current")
-        target = self._request.get("target")
-        if not current or not target:
+        if not self._options["archivers"]["extract"] or "target" not in self._request:
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
        if not self._options["archivers"]["extract"] or not target:
```

File: jupyter_elfinder/elfinder.py [hunk 19]
```diff
@@ -1360,34 +1440,29 @@ def __archive(self) -> None:
...
-            self.__content(cur_dir, False)
-            self._response["select"] = [self.__hash(archive_path)]
+            self._response["added"] = [self.__info(archive_path)]
         else:
             self._response["error"] = "Unable to create archive"
 
     def __extract(self) -> None:
         """Uncompress archive."""
-        current = self._request.get("current")
-        target = self._request.get("target")
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]: Restore this.

File: jupyter_elfinder/elfinder.py [hunk 20]
```diff
@@ -1360,34 +1440,29 @@ def __archive(self) -> None:
...
-        if not current or not target:
+        if not self._options["archivers"]["extract"] or "target" not in self._request:
             self._response["error"] = "Invalid parameters"
             return
-
-        cur_dir = self.__find_dir(self._request["current"], None)
-
-        if not cur_dir:
-            self._response["error"] = "File not found"
+        target = self._request["target"]
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
```

File: jupyter_elfinder/elfinder.py [hunk 21]
```diff
@@ -1360,34 +1440,29 @@ def __archive(self) -> None:
...
+        makedir = self._request.get("makedir")
+        if not target:
+            self._response["error"] = "Invalid parameters"
             return
```
Comment (MartinHjelmare) [inline jupyter_elfinder/elfinder.py]:
```suggestion
```

**Comment by oeway** [comment]
@MartinHjelmare I think I incorporated all your comments, thanks a lot!

**Comment by oeway** [comment]
Finally get the green check! Will merge now, thanks again.

**Pull request merged**